qiskit/primitives/containers/bindings_array.py

```diff
@@ -0,0 +1,357 @@
+    """Return ``val`` or ``val[..., None]``.
+
+    Args:
+        val: The array whose shape to standardize.
+        shape: The shape to standardize to.
+
+    Returns:
+        An array with one more dimension than ``len(shape)``, and whose leading dimensions match
+        ``shape``.
+
+    Raises:
+        ValueError: If the leading shape of ``val`` does not match the ``shape``.
+    """
+    if val.shape == shape:
+        val = val[..., None]
+    elif val.ndim - 1 != len(shape) or val.shape[:-1] != shape:
+        raise ValueError(f"Array with shape {val.shape} inconsistent with {shape}")
+    return val
+
+
+def _infer_shape(
+    vals: list[np.ndarray], kwvals: dict[tuple[Parameter, ...], np.ndarray]
+) -> tuple[int, ...]:
+    """Return a shape tuple that consistently defines the leading dimensions of all arrays.
+
+    Args:
+        vals: A list of arrays.
+        kwvals: A mapping from tuples to arrays, where the length of each tuple should match the
+            last dimension of the corresponding array.
+
+    Returns:
+        A shape tuple that matches the leading dimension of every array.
+
+    Raises:
+        ValueError: If this cannot be done unambiguously.
+    """
+    only_possible_shapes = None
+
+    def examine_array(*possible_shapes):
+        nonlocal only_possible_shapes
+        if only_possible_shapes is None:
+            only_possible_shapes = set(possible_shapes)
+        else:
+            only_possible_shapes.intersection_update(possible_shapes)
+
+    for parameters, val in kwvals.items():
+        if len(parameters) > 1:
+            # here, the last dimension _has_  to be over parameters
+            examine_array(val.shape[:-1])
+        elif val.shape == () or val.shape == (1,) or val.shape[-1] != 1:
+            # here, if the last dimension is not 1 or shape is () or (1,) then the shape is the shape
+            examine_array(val.shape)
+        else:
+            # here, the last dimension could be over parameters or not
+            examine_array(val.shape, val.shape[:-1])
+
+    if len(vals) == 1 and len(kwvals) == 0:
+        examine_array(vals[0].shape[:-1])
+    elif len(vals) == 0 and len(kwvals) == 0:
+        examine_array(())
+    else:
+        for val in vals:
+            # here, the last dimension could be over parameters or not
+            examine_array(val.shape, val.shape[:-1])
+
+    if len(only_possible_shapes) == 1:
+        return next(iter(only_possible_shapes))
+    elif len(only_possible_shapes) == 0:
+        raise ValueError("Could not find any consistent shape.")
+    raise ValueError("Could not unambiguously determine the intended shape; specify shape manually")
+
+
+def _format_key(key: tuple[Parameter | str, ...]):
+    return tuple(k.name if isinstance(k, Parameter) else k for k in key)
+
+
+BindingsArrayLike = Union[
+    BindingsArray,
+    NDArray,
+    "Mapping[Parameter, NDArray]",
+    "Sequence[NDArray]",
```

Comment:
`NDArray` might be relaxed to `ArrayLike`?
```suggestion
    ArrayLike,
    "Mapping[Parameter, ArrayLike]",
    "Sequence[ArrayLike]",
```

Reply:
probably